src/mongocxx/test/database.cpp

```diff
@@ -492,4 +492,95 @@ TEST_CASE("Database integration tests", "[database]") {
     }
 }
 
+// As C++11 lacks generic lambdas, and "ordinary" templates can't appear at block scope,
+// we'll have to define our helper for the serviceId tests here. This implementation would
+// be more straightforward in newer versions of C++ (C++14 and on allow generic lambdas),
+// see CXX-2350 (migration to more recent C++ standards); our C++17 optional<> implmentation
+// also has a few inconsistencies with the standard, which appear to vary across platforms:
+template <typename EventT>
+struct check_service_id {
+    const bool expect_service_id;
+
+    check_service_id(const bool expect_service_id) : expect_service_id(expect_service_id) {}
+
+    void operator()(const EventT& event) {
+        INFO("checking for service_id()")
+        CAPTURE(event.command_name(), expect_service_id);
+
+        auto service_id = event.service_id();
+
+        if (expect_service_id)
+            CHECK(service_id);
+        else
+            CHECK_FALSE(service_id);
+    }
+};
+
+TEST_CASE("serviceId presence depends on load-balancing mode") {
+    // Repeat this test case for a situation in which service_id should and should not be returned:
+    auto expect_service_id = GENERATE(true, false);
+
+    instance::current();
+
+    auto client_opts = test_util::add_test_server_api();
+
+    // Set Application Performance Monitoring options (APM):
+    mongocxx::options::apm apm_opts;
+    apm_opts.on_command_started(
+        check_service_id<mongocxx::events::command_started_event>{expect_service_id});
+    apm_opts.on_command_succeeded(
+        check_service_id<mongocxx::events::command_succeeded_event>{expect_service_id});
+    apm_opts.on_command_failed(
+        check_service_id<mongocxx::events::command_failed_event>{expect_service_id});
+
+    // Set up mocking for mongoc_apm_command_started_get_service_id:
```

Comment:
This applies to all three events. I suggest removing this comment since the comments on L544 and L560 note that mocks are being set.

Reply:
Given the high level of confusion that's possible for new-to-the-test-framework folks, I would prefer to leave these here. I agree it's somewhat redundant once it's clear what's going on, but like many other things in this test I think a bit of clarification might very well help the next reader, or one less familiar with the test harnesses. (Although if other reviewers are of the same opinion, I don't have super strong feelings about it.)

Reply:
Leaving the comment SGTM. It currently only notes `mongoc_apm_command_started_get_service_id` being mocked. Since `mongoc_apm_command_succeeded_get_service_id` and `mongoc_apm_command_failed_get_service_id` are also mocked, should this be edited to note all three?

Reply:
Ah! I see what you're saying! I've gone ahead and revised the comment. Derp!